File: uva online judge/10305 - Ordering Tasks.py
def read_test_case():
    n, m = (int(i) for i in input().split())
    if n == 0:
        return None
    adj_list = [[] for i in range(n)]
    for i in range(m):
        v, u = (int(i) - 1 for i in input().split())
        adj_list[v].append(u)
    return n, adj_list

def dfs(adj_list, v, visited, nodes):
    visited[v] = True
    for u in adj_list[v]:
        if not visited[u]:
            dfs(adj_list, u, visited, nodes)
    nodes.append(v)

def toposort(n, adj_list):
    nodes = []
    visited = [False] * n
    for v in range(n):
        if not visited[v]:
            dfs(adj_list, v, visited, nodes)
    nodes.reverse()
    return nodes

if __name__ == '__main__':
    for n, adj_list in iter(read_test_case, None):
        print(*(i + 1 for i in toposort(n, adj_list)))

File: uva online judge/10305 - Ordering Tasks.c
#include <stdio.h>
#define MAX_NODES 100

int adj_list[MAX_NODES][MAX_NODES + 1], visited[MAX_NODES], order[MAX_NODES];

int read_test_case(int *n, int adj_list[][MAX_NODES + 1]){
    int m, i, v, u;
    scanf("%d %d", n, &m);
    if(*n == 0)
        return 0;
    for(i = 0; i < *n; i++)
        adj_list[i][0] = 0;
    for(i = 0; i < m; i++){
        scanf("%d %d", &v, &u);
        adj_list[v - 1][++adj_list[v - 1][0]] = u - 1;
    }
    return 1;
}

void dfs(int n, int adj_list[][MAX_NODES + 1], int v, int visited[], int order[], int *idx){
    int i;
    visited[v] = 1;
    for(i = 1; i <= adj_list[v][0]; i++)
        if(!visited[adj_list[v][i]])
            dfs(n, adj_list, adj_list[v][i], visited, order, idx);
    order[(*idx)++] = v;
}

void toposort(int n, int adj_list[][MAX_NODES + 1], int visited[], int order[]){
    int v, idx = 0;
    for(v = 0; v < n; v++)
        visited[v] = 0;
    for(v = 0; v < n; v++)
        if(!visited[v])
            dfs(n, adj_list, v, visited, order, &idx);
}

int main(){
    int n, i;
    while(read_test_case(&n, adj_list)){
        toposort(n, adj_list, visited, order);
        for(i = n - 1; i >= 0; i--)
            printf("%d ", order[i] + 1);
        printf("\n");
    }
    
    return 0;
}